Threaded Context Manager

shannynalayna opened the pull request:
Adds the implementation for the threaded Context Manager, including a ContextManagerCore which holds an internal context manager which is used based on the messages received. Also adds a ContextManagerJoinHandle which is used to start the ContextManager thread and to create ContextManagers which hold a sender to send messages to the ContextManagerCore. 

Changes the references to the thread-safe Context Manager in the sync module used in the static execution adapter to the threaded version. Also changes the ContextLifecycle trait to be used by the Context Manager thread handler so the scheduler is using the ContextManagerJoinHandle and threaded ContextManager.

Commit:
Add initial ContextManagerCore struct

Adds initial ContextManagerCore struct to be used to manage the
threaded version of the ContextManager.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Add errors for ContextManagerCore

Adds errors to ContextManagerError module to handle Send and Receive
errors for the ContextManagerCore and handler.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Add interface for threaded Context Manager

Adds an interface for the threaded Context Manager which
communicates directly with the ContextManagerCore thread via
channels, which the threaded ContextManager interface holds
a sender to.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Add tests for the threaded ContextManager

Adds tests for Context Manager thread and handler.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Use threaded Context Manager in execution adapter

Changes references from the thread-safe Context Manager to the
Context Manager thread handler which utilizes the threaded
version of the Context Manager.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Alter ContextLifecycle api

Alters the ContextLifecycle api so that it may be used by the
threaded ContextManager, which forced a change in the return type
to a Result type which allows for error handling when used for the
threaded Context Manager.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Use ContextLifecycle for threaded ContextManager

Implements the ContextLifecycle trait for the Context Manager's
core handler. Also imports the ContextLifecycle trait in order
to allow to use the functions defined in the trait where the thread
handler is being used.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

peterschwarz commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -138,18 +138,20 @@ impl ContextManagerThread {
                 Err(err) => {
-                    error!("ContextManagerThread recv error: {}", err);
-                    break;
+                    return Err(ContextManagerThreadError::ThreadReceiveError(err));
```
Now that you have the `From` implementation, you could change this to `match self.thread_receiver.recv()?` and remove the need for the `Ok`/`Err` items in the match arms.

peterschwarz commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -138,18 +138,20 @@ impl ContextManagerThread {
             .spawn(move || {
-                self.run();
+                if let Err(err) = self.run() {
+                    error!("ContextManagerThread ended due to error: {}", err);
```
Generally, this error means all senders have been dropped, so it may mean that the parts of the system that send item to the background thread have also been shutdown and their resources cleaned up before a shutdown signal has been sent.

peterschwarz commented on libtransact/src/context/manager/thread/handler.rs:
```diff
@@ -0,0 +1,263 @@
+        }
+    }
+
+    pub fn shutdown(&mut self) -> Result<(), ContextManagerThreadError> {
```
Change this to `mut self`, so that the handler is consumed.

peterschwarz replied:
This should also be on a separate struct that is solely responsible for sending a shutdown message, since there can only be one join_handler, owned by the creator, but their can be multiple instances of the interface for context manager around the sender.

peterschwarz commented on libtransact/src/execution/adapter/static_adapter.rs:
```diff
@@ -21,7 +21,7 @@
 use std::thread;
 
-use crate::context::manager::sync::ContextManager;
+use crate::context::manager::thread::handler::ContextManagerThreadHandler;
```
Seeing it here, I'm not sure I like the name change.

peterschwarz commented on libtransact/src/context/mod.rs:
```diff
@@ -22,23 +22,27 @@ pub type ContextId = [u8; 16];
         context_id: &ContextId,
         transaction_id: &str,
-    ) -> Result<TransactionReceipt, ContextManagerError>;
+    ) -> Result<TransactionReceipt, ContextManagerThreadError>;
```
These should remain `ContextManagerError`, from the perspective of the trait caller.  In other words, the user of the trait shouldn't need to know anything about the underlying implementation.

Commit:
Add internal error for Context Manager thread

Adds an InternalError for the Context Manager to display errors
occurring within the Context Manager thread for external apis.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Add ContextManagerJoinHandle

Adds a handler for the ContextManagerCore to deal with
initiating and stopping the core's thread. Also creates
a new ContextManager which communicates with the core thread
via channels.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

peterschwarz commented on libtransact/src/context/manager/thread/core.rs:
```diff
@@ -246,3 +250,53 @@ impl ContextManagerCore {
+
+    /// Creates a new ContextManager and ContextManager thread.
+    ///
+    pub fn spawn_context_manager_thread(
```
this should be a function in the module, not part of this struct. I.e.

```rust
let (join_handle, context_manager) = transact::context::manager::thread::spawn_context_manager_thread(database);
```

Commit:
Update Sawtooth compat layer

Updates the Sawtooth compatability layer to use
the threaded version of the Context Manager. Also updates
the tests to use the threaded version.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

vaporos commented on libtransact/src/context/manager/thread/core.rs:
```diff
@@ -0,0 +1,308 @@
+
+/// Creates a new ContextManager and ContextManager thread.
+///
+pub fn spawn_context_manager_thread(
```
Exposing this low-level threading detail doesn't feel great from an API perspective.

peterschwarz replied:
This is effectively how the whole system is started.  It's analogous to `std::thread::spawn`.  What would you suggest here?

peterschwarz replied:
Perhaps a different name?

shannynalayna replied:
Could simply use spawn, but not sure since that still has threading connotations for the API

vaporos replied:
In the scheduler, we create the thread in ::new() and have a shutdown() method. Why not do that?

vaporos commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    /// Sets a key and value pair in the specified context's state.
+    ///
```
No need for the empty comment lines on these.

vaporos commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    /// Returns a list of key and value pairs from the specified context's state using the
+    ///     list of keys.
```
The reverse indent here feels awkward - is this actually a common rustdoc thing?

vaporos commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    fn drop_context(&mut self, _context_id: ContextId) {
+        unimplemented!();
```
Could this be implemented in this PR?

shannynalayna replied:
It certainly could be, I think before there was still a question on how exactly it would drop the contexts that I don't think has been resolved (At least from what I've heard)

vaporos commented on libtransact/src/context/manager/thread/mod.rs:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    #[test]
+    fn test_context_lifecycle() {
```
The test methods should have doc commentary.

Commit:
Add return type to drop_context

Adds a result return type to the drop_context function.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Implement drop_context in threaded Context Manager

Implements the drop_context manager message and handling for the
threaded ContextManager, although the internal ContextManager has
still not implemented the functionality for drop_context.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

peterschwarz commented on libtransact/src/context/manager/thread/core.rs:
```diff
@@ -68,6 +68,10 @@ pub enum ContextOperationMessage {
         state_id: String,
     },
+    DropContext {
+        handler_sender: Sender<ContextOperationResponse>,
```
I'm not sure if this actually needs a reply

peterschwarz replied:
Scratch that, there are some errors that can be returned.

peterschwarz commented on libtransact/src/context/manager/thread/core.rs:
```diff
@@ -159,6 +163,20 @@ impl ContextManagerCore {
+                    }
+                    Err(err) => {
+                        handler_sender.send(ContextOperationResponse::InvalidResult {
+                            error_message: err.description().to_string(),
```
This should probably just be `err.to_string()`
